programs/survival/bridge.sc

```diff
@@ -0,0 +1,49 @@
+global_bridge_item = 'golden_sword';
+
+__can_bridge(entity) -> 
+(
+	if (entity ~ 'gamemode' == 'adventure', return(false));
+	if (!(entity ~ 'sneaking'), return(false));
+	holds_mainhand = query(entity, 'holds', 'mainhand');
+	holds_offhand = query(entity, 'holds', 'offhand');
+	if (!holds_mainhand || !holds_offhand, return(false));
+	if (!(holds_mainhand:0 ~ global_bridge_item), return(false));
+	true
+);
+
+__get_direction_vector(face) ->
+(
+	if (face == 'up', return(l(0, 1, 0)));
+	if (face == 'down', return(l(0, -1, 0)));
+	if (face == 'north', return(l(0, 0, -1)));
+	if (face == 'south', return(l(0, 0, 1)));
+	if (face == 'east', return(l(1, 0, 0)));
+	if (face == 'west', return(l(-1, 0, 0)));
+	l(0, 0, 0)
+);
+
+__remove_offhand(player) ->
+(
+	holds_offhand = query(player, 'holds', 'offhand');
+	inventory_set(player, -1, holds_offhand:1 - 1)
+);
+
+__place_block(player, pos, dir, num) ->
+(
+	holds_offhand = query(player, 'holds', 'offhand');
+	if (!holds_offhand, return());
+	if (num <= 0, return());
+	l(x, y, z) = pos;
+	if(!place_item(holds_offhand:0, x, y, z), return());
+	if(player ~ 'gamemode' != 'creative', __remove_offhand(player));
+	schedule(5, '__place_block', player, pos + dir, dir, num - 1)
+);
+
+__on_player_right_clicks_block(player, item_tuple, hand, block, face, hitvec) ->
+(
+	if (hand != 'mainhand', return());
+	if(__can_bridge(player), (
+		dir = __get_direction_vector(face);
+		__place_block(player, pos(block) + dir, dir, 5)
```

Comment:
Really cool and simple application idea, but why just only 5 blocks?

Also quite dangerous when you try to pillar up

Reply:
I could raise the block limit, but I'm not sure what it should be. It could maybe even bridge until you run out of blocks. Also, I'm not quite sure how to fix pillaring up, the only idea I have is teleporting the player one block up if they're on top of the block being placed, although that might feel clunky.

Reply:
if its just for bridging out, just to make sure the offset on Y is 0, or the direction (face) is not `up`, or `down`. Also you can use inventory_find to get the block in the offhand and pull from the main inventory first - this will leave the offhand as the last resort, thus preventing replacing the stack after each 64 blocks. This may mean you could pillar up too far, but you can also enforce player sneaking to continue scheduling new events. If the player unshifts, that would stop building the bridge. Or just check the item in main hand, so you could run on it while its building. 

Throwing some ideas right there.

Reply:
I think that bridging up and down really comes in handy, so I think for bridging up the player will be teleported a block up if the are inside the block being placed. Also, I feel like pulling from the offhand acts as a nice cap to the bridging, because holding sneak feels a little bit annoying.